### PR TITLE
feat(groups): consolidate API for joining groups

### PR DESCRIPTION
**Pull request by hypeJunction**
Removes groups_join_group() and moves additional groups plugin logic
to an event handler.

BREAKING CHANGE
groups_join_group() has been removed: use ElggGroup::join() instead.

### CODE REVIEW

File: mod/groups/start.php [hunk 1]
```diff
@@ -680,41 +695,6 @@ function groups_get_invited_groups($user_guid, $return_guids = false, $options =
...
- * Join a user to a group, add river event, clean-up invitations
- *
- * @param ElggGroup $group
- * @param ElggUser  $user
- * @return bool
- */
-function groups_join_group($group, $user) {
-
-	// access ignore so user can be added to access collection of invisible group
-	$ia = elgg_set_ignore_access(TRUE);
```
Comment (jdalsem) [inline mod/groups/start.php]: do we need to tackle this inside group->join ??

Comment (hypeJunction) [inline mod/groups/start.php]: This shouldn't be there to start with. I looked through the stack and there is no reason for ignored access. If handlers need it, they can add it themselves. `$group` is already loaded and access hooks don't check access.

Comment (jdalsem) [inline mod/groups/start.php]: Reason is the hidden groups.... https://github.com/Elgg/Elgg/issues/1926 describes the issue... could you test if that is no longer an issue?

**Comment by hypeJunction** [comment]
We need to allow passing options to `ElggGroup::join()`, e.g. to disable river item creation - would make sense when the group is initially created and the owner is added as member.

File: mod/groups/start.php (hunk 1 above)
Comment (jdalsem) [inline mod/groups/start.php]: i tested the scenario described in the ticket... the problem is back... so there still needs to be done something special with the entity (maybe not as part of join) in the invite action

**Comment by hypeJunction** [comment]
This has been reworked, added some tests against #1926

File: engine/classes/ElggGroup.php [hunk 2]
```diff
@@ -205,32 +205,62 @@ public function isMember(\ElggUser $user = null) {
...
-		// event needs to be triggered while user is still member of group to have access to group acl
-		$params = array('group' => $this, 'user' => $user);
-		_elgg_services()->events->trigger('leave', 'group', $params);
+	public function leave(\ElggUser $user, array $params = []) {
+
+		// hook needs to be triggered while user is still member of group to have access to group acl
+		$params['group'] = $this;
+		$params['user'] = $user;
+		
+		if (_elgg_services()->events->hasHandler('leave', 'group')) {
```
Comment (jdalsem) [inline engine/classes/ElggGroup.php]: why not use `triggerDeprecated`

File: engine/classes/ElggGroup.php [hunk 3]
```diff
@@ -205,32 +205,62 @@ public function isMember(\ElggUser $user = null) {
...
-			$params = array('group' => $this, 'user' => $user);
+	public function join(\ElggUser $user, array $params = []) {
+
+		if (!add_entity_relationship($user->guid, 'member', $this->guid)) {
+			return false;
+		}
+
+		$params['group'] = $this;
+		$params['user'] = $user;
+		if (_elgg_services()->events->hasHandler('join', 'group')) {
```
Comment (jdalsem) [inline engine/classes/ElggGroup.php]: why not use `triggerDeprecated`?

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: mod/groups/start.php (hunk 1 above)
Comment (jdalsem) [inline mod/groups/start.php]: other than creating a test for the scenario... what did you change to fix this?

**Comment by jdalsem** [comment]
if you address my requested changes this can be merged

**Comment by mrclay** [comment]
Is this switching from an event to hook just to send params? Again wondering we we don't add params to events.

**Comment by hypeJunction** [comment]
Our notification system is based on events and expects a single event object. I don't think we can easily add params to events (nor should we). I think of Elgg events as a notification about an entity mutation, everything else is hooks. In this particular case, I would even remove the event/hook completely - it's a duplication of create, relationship event, and creates additional boilerplate.

File: mod/groups/start.php (hunk 1 above)
Comment (hypeJunction) [inline mod/groups/start.php]: The code reading the group_acl in the hook was wrapped in elgg_set_ignore_access()

Comment (hypeJunction) [inline mod/groups/start.php]: https://github.com/Elgg/Elgg/pull/10661/files#diff-d69a9c513b5e301164eb156de0de8db5R606

**Comment by jdalsem** [comment]
why was this closed? It was on the right track...

**Comment by hypeJunction** [comment]
I deleted my fork

**Comment by hypeJunction** [comment]
The branches have diverged too much. Should have been merged when it was still fresh. Not going to spend time recovering this.

**Pull request closed without merging**